libp2p/protocols/relay.nim

```diff
@@ -0,0 +1,491 @@
+    return
+  let
+    src = msg.srcPeer.get()
+    dst = msg.dstPeer.get()
+
+  # TODO: if r.acl # access control list?
+  #       and not r.acl.AllowHop(src.peerId, dst.peerId)
+  #         sendStatus(conn, RelayStatus.HopCantSpeakRelay)
+
+  r.hopCount.inc(src.peerId)
+  r.hopCount.inc(dst.peerId)
+  defer:
+    r.hopCount.inc(src.peerId, -1)
+    r.hopCount.inc(dst.peerId, -1)
+
+  if r.hopCount[src.peerId] > r.maxCircuitPerPeer:
+    trace "refusing connection; too many connection from src", src, dst
+    await sendStatus(conn, RelayStatus.HopCantSpeakRelay)
+    return
+
+  if r.hopCount[dst.peerId] > r.maxCircuitPerPeer:
+    trace "refusing connection; too many connection to dst", src, dst
+    await sendStatus(conn, RelayStatus.HopCantSpeakRelay)
+    return
+
+  let connDst = try:
+    await r.switch.dial(dst.peerId, @[RelayCodec])
+  except CancelledError as exc:
+    raise exc
+  except CatchableError as exc:
+    trace "error opening relay stream", dst, exc=exc.msg
+    await sendStatus(conn, RelayStatus.HopCantDialDst)
+    return
+  defer:
+    await connDst.close()
+
+  let msgToSend = RelayMessage(
+    msgType: some(RelayType.Stop),
+    srcPeer: some(src),
+    dstPeer: some(dst),
+    status: none(RelayStatus))
+  var pb = encodeMsg(msgToSend)
+  try:
+    await connDst.writeLp(pb.buffer)
+  except CancelledError as exc:
+    raise exc
+  except CatchableError as exc:
+    trace "error writing stop handshake", exc=exc.msg
+    await sendStatus(conn, RelayStatus.HopCantOpenDstStream)
+    return
+
+  let msgRcvFromDstOpt = try:
+    decodeMsg(await connDst.readLp(r.msgSize))
+  except CancelledError as exc:
+    raise exc
+  except CatchableError as exc:
+    trace "error reading stop response", exc=exc.msg
+    await sendStatus(conn, RelayStatus.HopCantOpenDstStream)
+    return
+
+  if msgRcvFromDstOpt.isNone:
+    trace "error reading stop response", msg = msgRcvFromDstOpt
+    await sendStatus(conn, RelayStatus.HopCantOpenDstStream)
+    return
+
+  let msgRcvFromDst = msgRcvFromDstOpt.get()
+  if msgRcvFromDst.msgType.isNone or msgRcvFromDst.msgType.get() != RelayType.Status:
+    trace "unexcepted relay stop response", msgType = msgRcvFromDst.msgType
+    await sendStatus(conn, RelayStatus.HopCantOpenDstStream)
+    return
+
+  if msgRcvFromDst.status.isNone or msgRcvFromDst.status.get() != RelayStatus.Success:
+    trace "relay stop failure", status=msgRcvFromDst.status
+    await sendStatus(conn, RelayStatus.HopCantOpenDstStream)
+    return
+
+  try:
+    await sendStatus(conn, RelayStatus.Success)
+  except CancelledError as exc:
+    raise exc
+  except CatchableError as exc:
+    trace "error writing relay response", exc=exc.msg
+    return
+
+  trace "relaying connection", src, dst
+
+  # TODO: Eventually create a similar procedure somewhere in the chronos library
+  proc bridge(conn: Connection, connDst: Connection) {.async.} =
+    var
+      bufSrcToDst: array[4096, byte]
```

Comment:
I would use the `{.noInit.}` pragma here and then `zeroMem` every time I finished using the buffer in the loop.

Reply:
Not sure I understand the issue here, this code can't leak any data unless `read` returns a wrong value, which would be catastrophic anyway
And even if a bit of the buffer was somehow reused, that would be in the same relayed connection, so it would only resend some old data to the same peer

And even if all of that happened, since the connection from A to B through R is secured anyway, you wouldn't be able to extract any useful data? R could post the entire exchange on Facebook and nothing bad would happen

The only thing we have to be careful about is to not leak our internal memory, so we must be careful to zero mem the buffers before any usage (which is done implicitly right now)

Reply:
After a little thought and more deeper inspection on how pipes work kernel-side (with a buffer, more or less the same thing I did) I must agree with @Menduist, there's not really a better solution.

A `const bufferSize = 4096` or `bufSrcToDst.high + 1` are good ways to avoid the repeat of the 4096 that could be problematic if someone updates this section of the code.

And I understand the use of the `zeroMem` at the end of the loop, even though, like Menduist pointed out, the connection is secured so it's a bit superfluous, but we never know what can happen. The `{.noInit.}` pragma on another hand is less effective to secure this part, if the code fails during the first loop, it could be dangerous with the buffer uninitialized.